File: 05-cypress-tests/cypress/pages/saucedemo/InventoryPage.js
// ============================================================
// SauceDemo — Inventory (Products) Page Object
//
// The product grid shown after login. Covers item listing, the
// sort dropdown, add-to-cart, and the cart badge / link.
// ============================================================

class InventoryPage {
  // ---- Selectors ----------------------------------------------
  elements = {
    inventoryItems:  () => cy.get('.inventory_item'),
    itemNames:       () => cy.get('.inventory_item_name'),
    itemPrices:      () => cy.get('.inventory_item_price'),
    sortDropdown:    () => cy.get('[data-test="product-sort-container"]'),
    cartBadge:       () => cy.get('.shopping_cart_badge'),
    cartLink:        () => cy.get('.shopping_cart_link'),

    // Add/remove buttons are keyed by a slugified product name, e.g.
    // [data-test="add-to-cart-sauce-labs-backpack"].
    addToCartButton: (slug) => cy.get(`[data-test="add-to-cart-${slug}"]`),
    removeButton:    (slug) => cy.get(`[data-test="remove-${slug}"]`),

    // Continue-shopping button on the cart page — used to navigate back
    // to the inventory IN-APP (no server request).
    continueShopping: () => cy.get('[data-test="continue-shopping"]'),
  }

  // ---- Actions ------------------------------------------------

  /**
   * Reset to a clean, empty-cart inventory between tests that share one
   * logged-in session (the testIsolation:false checkout spec).
   *
   * Why this shape (clear localStorage + cart-link → continue-shopping):
   *   - SauceDemo serves /inventory.html with an HTTP 404 status (it's a
   *     History-API SPA). cy.visit / cy.reload on that path makes Cypress
   *     choke on the resulting DOMException intermittently — so we never
   *     re-navigate by URL.
   *   - A fresh UI login per test (8×) trips SauceDemo's CDN rate-limit
   *     even from CI — so we log in once and reset in-app.
   *   - The burger menu works too, but its slide-out animation makes the
   *     reset/all-items clicks flaky.
   * Clearing the `cart-contents` localStorage key empties the cart; the
   * two in-app navigations (cart link, then "Continue Shopping") force a
   * pushState re-render of the inventory from the clean state, with no
   * network round-trip and no animation timing to fight.
   */
  resetToCleanInventory() {
    cy.window().then((w) => w.localStorage.removeItem('cart-contents'))
    this.elements.cartLink().click()
    this.elements.continueShopping().click()
    this.elements.inventoryItems().should('have.length', 6)
    this.elements.cartBadge().should('not.exist')
    return this
  }

  /**
   * Select a sort option by its <option> value:
   *   az  = Name A→Z   |  za = Name Z→A
   *   lohi = Price low→high  |  hilo = Price high→low
   */
  sortBy(value) {
    this.elements.sortDropdown().select(value)
    return this
  }

  addToCart(slug) {
    this.elements.addToCartButton(slug).click()
    return this
  }

  openCart() {
    this.elements.cartLink().click()
    return this
  }

  // ---- Assertions / Queries -----------------------------------

  assertItemCount(expected) {
    this.elements.inventoryItems().should('have.length', expected)
    return this
  }

  assertCartBadgeCount(expected) {
    this.elements.cartBadge().should('contain', String(expected))
    return this
  }

  /**
   * Assert prices are sorted ascending. Reads the rendered price
   * text, strips the $, and verifies the array is non-decreasing.
   */
  assertPricesAscending() {
    this.elements.itemPrices().then(($els) => {
      const prices = [...$els].map((el) => parseFloat(el.innerText.replace('$', '')))
      const sorted = [...prices].sort((a, b) => a - b)
      expect(prices, 'prices should be ascending').to.deep.equal(sorted)
    })
    return this
  }

  /**
   * Assert product names are in reverse-alphabetical (Z→A) order.
   */
  assertNamesDescending() {
    this.elements.itemNames().then(($els) => {
      const names = [...$els].map((el) => el.innerText)
      const sorted = [...names].sort().reverse()
      expect(names, 'names should be Z→A').to.deep.equal(sorted)
    })
    return this
  }
}

export default new InventoryPage()

File: 07-selenium-java/src/main/java/pages/saucedemo/InventoryPage.java
package pages.saucedemo;

import org.openqa.selenium.By;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;

import java.time.Duration;
import java.util.List;
import java.util.stream.Collectors;

/**
 * SauceDemo — Inventory (Products) Page Object.
 *
 * <p>The product grid shown after login: item listing, the sort dropdown,
 * add-to-cart, and the cart badge / link. Mirrors the Cypress and Playwright
 * {@code InventoryPage} objects.
 */
public class InventoryPage {

    private final WebDriver driver;
    private final WebDriverWait wait;

    // ---- Selectors ------------------------------------------------
    private static final By INVENTORY_ITEMS = By.className("inventory_item");
    private static final By ITEM_NAMES       = By.className("inventory_item_name");
    private static final By ITEM_PRICES      = By.className("inventory_item_price");
    private static final By SORT_DROPDOWN    = By.cssSelector("[data-test=\"product-sort-container\"]");
    private static final By CART_BADGE = By.className("shopping_cart_badge");
    private static final By CART_LINK  = By.className("shopping_cart_link");

    // Add/remove buttons are keyed by a slugified product name, e.g.
    // [data-test="add-to-cart-sauce-labs-backpack"] / [data-test="remove-..."].
    private static By addToCartButton(String slug) {
        return By.cssSelector("[data-test=\"add-to-cart-" + slug + "\"]");
    }

    private static By removeButton(String slug) {
        return By.cssSelector("[data-test=\"remove-" + slug + "\"]");
    }

    public InventoryPage(WebDriver driver) {
        this.driver = driver;
        this.wait   = new WebDriverWait(driver, Duration.ofSeconds(10));
    }

    // ---- Actions --------------------------------------------------

    /**
     * Select a sort option by its &lt;option&gt; value:
     * {@code az} = Name A→Z, {@code za} = Name Z→A,
     * {@code lohi} = Price low→high, {@code hilo} = Price high→low.
     */
    public InventoryPage sortBy(String value) {
        new Select(driver.findElement(SORT_DROPDOWN)).selectByValue(value);
        return this;
    }

    /**
     * Add a product to the cart, confirming the add registered (the button
     * flips to "Remove"), and retrying if the click was lost.
     *
     * <p>SauceDemo's React buttons can drop a click before their handler is
     * fully wired under load, so we verify the outcome and re-click instead of
     * trusting a single fire — which also keeps the cart-badge read reliable.
     */
    public InventoryPage addToCart(String slug) {
        By removeBtn = removeButton(slug);
        for (int attempt = 1; attempt <= 3; attempt++) {
            if (!driver.findElements(removeBtn).isEmpty()) {
                return this; // already in the cart
            }
            wait.until(ExpectedConditions.elementToBeClickable(addToCartButton(slug))).click();
            try {
                new WebDriverWait(driver, Duration.ofSeconds(3))
                        .until(ExpectedConditions.visibilityOfElementLocated(removeBtn));
                return this;
            } catch (TimeoutException dropped) {
                // click was lost before React wired up — retry
            }
        }
        throw new IllegalStateException("Add-to-cart did not register for: " + slug);
    }

    public InventoryPage openCart() {
        driver.findElement(CART_LINK).click();
        wait.until(ExpectedConditions.urlContains("/cart.html"));
        return this;
    }

    // ---- Assertions / Queries -------------------------------------

    public int itemCount() {
        return driver.findElements(INVENTORY_ITEMS).size();
    }

    /** Cart-badge count, or 0 when the badge is absent (empty cart). */
    public int cartBadgeCount() {
        List<WebElement> badges = driver.findElements(CART_BADGE);
        return badges.isEmpty() ? 0 : Integer.parseInt(badges.get(0).getText().trim());
    }

    /** Rendered prices as doubles, in display order (for sort assertions). */
    public List<Double> prices() {
        return driver.findElements(ITEM_PRICES).stream()
                .map(el -> Double.parseDouble(el.getText().replace("$", "")))
                .collect(Collectors.toList());
    }

    /** Rendered product names, in display order (for sort assertions). */
    public List<String> names() {
        return driver.findElements(ITEM_NAMES).stream()
                .map(WebElement::getText)
                .collect(Collectors.toList());
    }
}
